### PR TITLE
MNTOR-1937: Autocomplete location cronjob

### PR DESCRIPTION
**Pull request by mansaj**
# References: 
Jira: MNTOR-1937

# Description
Move autocomplete location json file fetching/bundling process to a cronjob. 
* The job will execute daily
* The job will create the json file and upload it to s3
* Another script added to simply fetch the file from s3 and put it in the root directory as a part of the build step
* Build time / e2e testing should take significantly less time

# Screenshot (if applicable)

Not applicable.

# How to test

# Checklist (Definition of Done)
- [ ] Localization strings (if needed) have been added.
- [ ] Commits in this PR are minimal and [have descriptive commit messages](https://chris.beams.io/posts/git-commit/).
- [ ] I've added or updated the relevant sections in readme and/or code comments
- [ ] I've added a unit test to test for potential regressions of this bug.
- [ ] Product Owner accepted the User Story (demo of functionality completed) or waived the privilege.
- [ ] All acceptance criteria are met.
- [ ] Jira ticket has been updated (if needed) to match changes made during the development process.
- [ ] Jira ticket has been updated (if needed) with suggestions for QA when this PR is deployed to stage.

### CODE REVIEW

File: src/scripts/uploadAutoCompleteLocations.js [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+const DATA_COUNTRY_CODE = "US";
+const LOCATIONS_DATA_FILE = "locationAutocompleteData.json";
+const FETCH_REMOTE_DATASETS = true;
+const CLEANUP_TMP_DATA_AFTER_FINISHED = true;
+const SENTRY_SLUG = "cron-create-location-autocomplete"
+
+Sentry.init({
+  dsn: process.env.SENTRY_DSN,
+  tracesSampleRate: 1.0,
+});
```
Comment (pdehaan) [inline src/scripts/uploadAutoCompleteLocations.js:42]: Do we need to check if `process.env.SENTRY_DSN` is defined?
Or more specifically, skip Sentry logging if this is running in `APP_ENV:"local"`?

File: src/scripts/uploadAutoCompleteLocations.js [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+
+        relevantLocations.push({
+          id: geonameId,
+          // switch names if an alternate name is the preferred location name
+          name: preferredName ? preferredName.name : name,
+          stateCode: admin1Code,
+          countryCode: "USA",
+          featureClass,
+          featureCode,
+          population,
```
Comment (pdehaan) [inline src/scripts/uploadAutoCompleteLocations.js:275]: Not sure if it matters, but looks like `population` is still a string from the TSV file, not sure where/if it's being used downstream and if we want to cast it to Number:

```js
{ "id":"4905687",
  "name":"Peoria",
  "stateCode":"IL",
  "countryCode":"USA",
  "featureClass":"P",
  "featureCode":"PPLA2",
  "population":"115070" }
```

File: src/scripts/getAutoCompleteLocations.js [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+import "dotenv/config"
+
+const stream = createWriteStream(resolve('./locationAutocompleteData.json'));
+try {
+  const fetchUrl = `https://s3.amazonaws.com/${process.env.S3_BUCKET}/autocomplete/locationAutocompleteData.json`;
+  console.debug({ fetchUrl })
+  const { body } = await fetch(fetchUrl);
+  await finished(Readable.fromWeb(body).pipe(stream));
+} catch (e) {
+  console.error(e)
```
Comment (pdehaan) [inline src/scripts/getAutoCompleteLocations.js]: Should this set a non-zero `process.exitCode`?

File: src/scripts/getAutoCompleteLocations.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import { resolve } from 'node:path'
+import { createWriteStream } from 'fs';
+import { Readable } from 'stream';
+import { finished } from 'stream/promises';
```
Comment (pdehaan) [inline src/scripts/getAutoCompleteLocations.js]: nit: should we prefix these w/ `node:` similar to L1?

File: src/scripts/getAutoCompleteLocations.js
Comment (pdehaan) [inline src/scripts/getAutoCompleteLocations.js:1]: nit: Add copyrght header?

File: src/scripts/getAutoCompleteLocations.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+import "dotenv/config"
+
+const dataPath = "./locationAutocompleteData.json";
+
+if (!existsSync(dataPath)) {
+  const stream = createWriteStream(dataPath);
+  try {
+    const fetchUrl = `https://s3.amazonaws.com/${process.env.S3_BUCKET}/autocomplete/locationAutocompleteData.json`;
+    console.debug({ fetchUrl })
```
Comment (pdehaan) [inline src/scripts/getAutoCompleteLocations.js:12]: ✂️

File: src/scripts/getAutoCompleteLocations.js [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+    const { body } = await fetch(fetchUrl);
+    await finished(Readable.fromWeb(body).pipe(stream));
+  } catch (e) {
+    console.error(e)
+  }
+} else {
+  console.log('file already exists: ', dataPath)
+}
```
Comment (pdehaan) [inline src/scripts/getAutoCompleteLocations.js:20]: nit: add trailing line break

File: src/scripts/uploadAutoCompleteLocations.js [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+    `-> ${totalCount}/${currentCount + 1} (${progress}%) \r`
+  );
+}
+
+function writeFromRemoteFile({
+  url,
+  writeStream,
+}) {
+  return new Promise((resolve, reject) => {
+    https.get(url, (res) => {
```
Comment (pdehaan) [inline src/scripts/uploadAutoCompleteLocations.js:76]: personal Q: Is there a benefit to using `https.get` and streams versus just `fetch()`?
or is it mainly just for memory concerns if these are big data files?

Comment (rhelmer) [reply src/scripts/uploadAutoCompleteLocations.js:76]: AFAIK `fetch` is still experimental in Node, but it's in pretty common use.

**Review by rhelmer — APPROVED** [review]
Thanks for doing this, lgtm overall - see the note about passing `APP_ENV` to Sentry, we need to do that explicitly AFAIK or it falls back to `NODE_ENV`.

**Pull request merged**